homeassistant/components/google_assistant/trait.py

```diff
@@ -1463,3 +1478,179 @@ def _verify_ack_challenge(data, state, challenge):
         return
     if not challenge or not challenge.get("ack"):
         raise ChallengeNeeded(CHALLENGE_ACK_NEEDED)
+
+
+MEDIA_COMMAND_SUPPORT_MAPPING = {
+    COMMAND_MEDIA_NEXT: media_player.SUPPORT_NEXT_TRACK,
+    COMMAND_MEDIA_PAUSE: media_player.SUPPORT_PAUSE,
+    COMMAND_MEDIA_PREVIOUS: media_player.SUPPORT_PREVIOUS_TRACK,
+    COMMAND_MEDIA_RESUME: media_player.SUPPORT_PLAY,
+    COMMAND_MEDIA_SEEK_RELATIVE: media_player.SUPPORT_SEEK,
+    COMMAND_MEDIA_SEEK_TO_POSITION: media_player.SUPPORT_SEEK,
+    COMMAND_MEDIA_SHUFFLE: media_player.SUPPORT_SHUFFLE_SET,
+    COMMAND_MEDIA_STOP: media_player.SUPPORT_STOP,
+}
+
+MEDIA_COMMAND_ATTRIBUTES = {
+    COMMAND_MEDIA_NEXT: "NEXT",
+    COMMAND_MEDIA_PAUSE: "PAUSE",
+    COMMAND_MEDIA_PREVIOUS: "PREVIOUS",
+    COMMAND_MEDIA_RESUME: "RESUME",
+    COMMAND_MEDIA_SEEK_RELATIVE: "SEEK_RELATIVE",
+    COMMAND_MEDIA_SEEK_TO_POSITION: "SEEK_TO_POSITION",
+    COMMAND_MEDIA_SHUFFLE: "SHUFFLE",
+    COMMAND_MEDIA_STOP: "STOP",
+}
+
+
+@register_trait
+class TransportControlTrait(_Trait):
+    """Trait to control media playback.
+
+    https://developers.google.com/actions/smarthome/traits/transportcontrol
+    """
+
+    name = TRAIT_TRANSPORT_CONTROL
+    commands = [
+        COMMAND_MEDIA_NEXT,
+        COMMAND_MEDIA_PAUSE,
+        COMMAND_MEDIA_PREVIOUS,
+        COMMAND_MEDIA_RESUME,
+        COMMAND_MEDIA_SEEK_RELATIVE,
+        COMMAND_MEDIA_SEEK_TO_POSITION,
+        COMMAND_MEDIA_SHUFFLE,
+        COMMAND_MEDIA_STOP,
+    ]
+
+    @staticmethod
+    def supported(domain, features, device_class):
+        """Test if state is supported."""
+        if domain == media_player.DOMAIN:
+            for feature in MEDIA_COMMAND_SUPPORT_MAPPING.values():
+                if features & feature:
+                    return True
+
+        return False
+
+    def sync_attributes(self):
+        """Return opening direction."""
+        response = {}
+
+        if self.state.domain == media_player.DOMAIN:
+            features = self.state.attributes.get(ATTR_SUPPORTED_FEATURES, 0)
+
+            support = []
+            for command, feature in MEDIA_COMMAND_SUPPORT_MAPPING.items():
+                if features & feature:
+                    support.append(MEDIA_COMMAND_ATTRIBUTES[command])
+            response["transportControlSupportedCommands"] = support
+
+        return response
+
+    def query_attributes(self):
+        """Return the attributes of this trait for this entity."""
+
+        return {}
+
+    async def execute(self, command, data, params, challenge):
+        """Execute a media command."""
+
+        service_attrs = {ATTR_ENTITY_ID: self.state.entity_id}
+
+        if command == COMMAND_MEDIA_SEEK_RELATIVE:
+            service = media_player.SERVICE_MEDIA_SEEK
+
+            rel_position = params["relativePositionMs"] / 1000
+            position = self.state.attributes.get(media_player.ATTR_MEDIA_POSITION, 0)
```

Comment:
You need to take into account ATTR_MEDIA_POSITION_UPDATED_AT to support relative seek. If state is playing, you need to append the delay from now.

Reply:
Looks good. One thing I don't know here is the behavior of this when paused.. I have checked if the time is exposed then. If it is, seeking will end up wrong.